geoopt/optim/rwolfe.py

```diff
@@ -0,0 +1,280 @@
+                manifold = point.manifold
+            else:  # Use euclidean manifold
+                manifold = self._default_manifold
+
+            grad = state["grad"]
+
+            # compute retract and transport in gradient direction
+            new_point, grad_retr = manifold.retr_transp(point, -step_size * grad, grad)
+            with torch.no_grad():
+                point.copy_(new_point)
+
+            state["grad_retr"] = grad_retr
+
+        # recompute loss at new point
+        phi = self.closure()
+
+        # Store new gradients in state
+        for point in self._params:
+            grad = point.grad
+            if grad is None:
+                continue
+            if isinstance(point, (ManifoldParameter, ManifoldTensor)):
+                manifold = point.manifold
+            else:  # Use euclidean manifold
+                manifold = self._default_manifold
+
+            state = self.state[point]
+
+            # project gradient onto tangent space
+            grad = manifold.egrad2rgrad(point, grad)
+
+            state["new_grad"] = grad
+
+            state["der_phi"] = torch.sum(
+                manifold.inner(point, -grad, state["grad_retr"])
+            )
+
+        # roll back parameters to before step
+        with torch.no_grad():
+            for point, old_point in zip(self._params, param_copy):
+                point.copy_(old_point)
+
+        self._step_size_dic[step_size] = phi
+        return phi
+
+    def derphi_(self, step_size):
+        """Compute derivative of phi.
+
+        The derivative of phi is given by computing inner
+        product between all tensor gradients at target point and those at source point.
+        The source gradients are transported to the target point, and both gradients are
+        projected.
+        """
+
+        # Call phi_ to compute gradients; Does nothing if phi_ was
+        # already called with this stepsize during this step
+        self.phi_(step_size)
+
+        derphi = 0
+        for point in self._params:
+            state = self.state[point]
+            if "der_phi" not in state:
+                continue
+
+            derphi += torch.sum(state["der_phi"])
+
+        return derphi
+
+    def init_loss(self):
+        """Compute loss and gradients at start of line search."""
+
+        loss = self.closure()
+        derphi0 = 0
+        self._step_size_dic = dict()
+
+        for point in self._params:
+            grad = point.grad
+            if grad is None:
+                continue
+            if isinstance(point, (ManifoldParameter, ManifoldTensor)):
+                manifold = point.manifold
+            else:  # Use euclidean manifold
+                manifold = self._default_manifold
+
+            state = self.state[point]
+            # project gradient onto tangent space
+            grad = manifold.egrad2rgrad(point, grad)
+            grad_norm = torch.sum(manifold.norm(point, grad)).item()
```

Comment:
Norm here is computed incorrect, I guess you meant
```suggestion
            grad_norm = torch.sum(manifold.inner(point, grad)).pow(.5).item()
```

Reply:
> Norm here is computed incorrect, I guess you meant

what is the difference between the two? Shouldn't manifold.norm do exactly this?